migrations/sqlite/20240814131142_auth_rebac_tables.sql

```diff
@@ -0,0 +1,28 @@
+CREATE TABLE auth_rebac_properties
+(
+    entity_type    VARCHAR(25)  NOT NULL,
+    entity_id      VARCHAR(100) NOT NULL,
+    property_name  VARCHAR(50)  NOT NULL,
+    property_value VARCHAR(50)  NOT NULL
+);
+
+CREATE INDEX idx_uniq_entity
+    ON auth_rebac_properties (entity_type, entity_id);
+
+CREATE UNIQUE INDEX idx_uniq_entity_property_name
+    ON auth_rebac_properties (entity_type, entity_id, property_name);
+
+------------------------------------------------------------------------------------------------------------------------
+
+CREATE TABLE auth_rebac_relations
+(
+    subject_entity_type VARCHAR(25)  NOT NULL,
+    subject_entity_id   VARCHAR(100) NOT NULL,
+    relationship        VARCHAR(50)  NOT NULL,
+    object_entity_type  VARCHAR(25)  NOT NULL,
+    object_entity_id    VARCHAR(100) NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_uniq_row
+    ON auth_rebac_relations (subject_entity_type, subject_entity_id, relationship, object_entity_type,
+                             object_entity_id);
```

Comment:
Earlier, I added more indexes:

```sql
CREATE INDEX idx_subject_entity
    ON auth_rebac_relations (subject_entity_type, subject_entity_id);

CREATE INDEX idx_subject_entity_relationship
    ON auth_rebac_relations (subject_entity_type, subject_entity_id, relationship);

CREATE INDEX idx_subject_entity_object_entity
    ON auth_rebac_relations (subject_entity_type, subject_entity_id, object_entity_type, object_entity_id);
```


But then I found out that they are not needed, since the queries all go through a unique index:
```sql
EXPLAIN QUERY PLAN
SELECT *
FROM auth_rebac_relations
WHERE subject_entity_type = 'account'
  AND subject_entity_id = 'user2';
```

| id | parent | notused | detail |
| :--- | :--- | :--- | :--- |
| 2 | 0 | 0 | SEARCH auth\_rebac\_relations USING COVERING INDEX idx\_uniq\_row \(subject\_entity\_type=? AND subject\_entity\_id=?\) |

```
EXPLAIN QUERY PLAN
SELECT *
FROM auth_rebac_relations
WHERE subject_entity_type = 'account'
  AND subject_entity_id = 'user2'
  AND relationship = 'reader';
```

| id | parent | notused | detail |
| :--- | :--- | :--- | :--- |
| 2 | 0 | 0 | SEARCH auth\_rebac\_relations USING COVERING INDEX idx\_uniq\_row \(subject\_entity\_type=? AND subject\_entity\_id=? AND relationship=?\) |

```sql
EXPLAIN QUERY PLAN
SELECT *
FROM auth_rebac_relations
WHERE subject_entity_type = 'account'
  AND subject_entity_id = 'user2'
  AND object_entity_type = 'database'
  AND object_entity_id = 'ds2';
```

| id | parent | notused | detail |
| :--- | :--- | :--- | :--- |
| 2 | 0 | 0 | SEARCH auth\_rebac\_relations USING COVERING INDEX idx\_uniq\_row \(subject\_entity\_type=? AND subject\_entity\_id=?\) |


`auth_rebac_relations` table content:
| subject\_entity\_type | subject\_entity\_id | relationship | object\_entity\_type | object\_entity\_id |
| :--- | :--- | :--- | :--- | :--- |
| account | user1 | reader | dataset | ds1 |
| account | user2 | reader | dataset | ds1 |
| account | user2 | reader | dataset | ds2 |
| account | user2 | reader | dataset | ds3 |
| account | user2 | reader | dataset | ds4 |